### PR TITLE
Dr2 1497 add lambda to call preservica client for version

### PR DESCRIPTION
**Pull request by techncl**

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        required: true
+
+permissions:
+  id-token: write
+  contents: write
+jobs:
+  deploy:
+    uses: nationalarchives/dr2-github-actions/.github/workflows/lambda_deploy.yml@main
+    with:
+      lambda-name: ${{ github.event.inputs.environment }}-dr2-get-latest-preservica-version
```
Comment (MancunianSam) [inline .github/workflows/deploy.yml]: There's the bug in the template again, you need to take the -dr2 out of here.

File: build.sbt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      awsDynamoDbClient,
+      dynamoFormatters,
+      log4jSlf4j,
+      log4jCore,
+      log4jTemplateJson,
+      lambdaCore,
+      awsSnsClient,
+      lambdaJavaEvents,
+      mockitoScala,
+      mockitoScalaTest,
```
Comment (MancunianSam) [inline build.sbt]: Can you put the to mockitoScala dependencies as `% Test` please.

File: project/Dependencies.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import sbt._
+object Dependencies {
+  private val mockitoScalaVersion = "1.17.30"
+  lazy val pureConfigVersion = "0.17.5"
+  lazy val logbackVersion = "2.22.1"
+  private val daAwsClientsVersion = "0.1.36"
```
Comment (MancunianSam) [inline project/Dependencies.scala]: Could we have these all either `lazy val` or `private lazy val` or `private val` just to keep things tidy.

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import uk.gov.nationalarchives.Lambda._
+import uk.gov.nationalarchives.dp.client.EntityClient
+import uk.gov.nationalarchives.dp.client.fs2.Fs2Client
+import upickle.default._
+
+import java.util.UUID
+
+class Lambda extends RequestHandler[ScheduledEvent, Unit] {
+
+  implicit val inputReader: Reader[Input] = macroR[Input]
```
Comment (MancunianSam) [inline src/main/scala/uk/gov/nationalarchives/Lambda.scala]: This `Input` case class isn't being used anywhere so you can delete this line, the case class declaration and the upickle imports as well.

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        List(PartitionKey("DR2PreservicaVersion")),
+        config.currentPreservicaVersionTableName
+      )
+      _ <- log("Retrieved the version of Preservica that we are using")
+
+      entitiesClient <- entitiesClientIO
+      latestPreservicaVersion <- entitiesClient.getPreservicaNamespaceVersion(lowImpactEndpoint)
+      _ <- log("Retrieved version of Preservica that Preservica are using")
+
+      versionWeAreUsing = versionWeAreUsingResponses.head.version
```
Comment (MancunianSam) [inline src/main/scala/uk/gov/nationalarchives/Lambda.scala]: This will fail if the version is missing from Dynamo, it's unlikely but we may as well cover it.
Something like 
```scala
versionWeAreUsing <- IO.fromOption(versionWeAreUsingResponses.headOption.map(_.version))(new RuntimeException("Version not found"))
```
Can you add a test in as well to simulate what happens if Dynamo returns an empty list.

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    for {
+      config <- configIo
+      logCtx: Map[String, String] = Map("endpointToCheck" -> lowImpactEndpoint)
+      log = logger.info(logCtx)(_)
+
+      versionWeAreUsingResponses <- dADynamoDBClient.getItems[GetDr2PreservicaVersionResponse, PartitionKey](
+        List(PartitionKey("DR2PreservicaVersion")),
+        config.currentPreservicaVersionTableName
+      )
+      _ <- log("Retrieved the version of Preservica that we are using")
```
Comment (MancunianSam) [inline src/main/scala/uk/gov/nationalarchives/Lambda.scala]: If you add the `versionWeAreUsing <- IO.fromOption(versionWeAreUsingResponses.headOption.map(_.version))(new RuntimeException("Version not found"))` line above here, you can log the version that's been retrieved from Dynamo.

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+      entitiesClient <- entitiesClientIO
+      latestPreservicaVersion <- entitiesClient.getPreservicaNamespaceVersion(lowImpactEndpoint)
+      _ <- log("Retrieved version of Preservica that Preservica are using")
```
Comment (MancunianSam) [inline src/main/scala/uk/gov/nationalarchives/Lambda.scala]: Is it worth logging the version we get from Preservica here?

File: src/main/scala/uk/gov/nationalarchives/Lambda.scala [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  private case class Config(
+      demoApiUrl: String,
+      secretName: String,
+      snsArn: String,
+      currentPreservicaVersionTableName: String
+  )
+
+  case class PartitionKey(id: String)
+
+  sealed trait PreservicaVersion {
```
Comment (MancunianSam) [inline src/main/scala/uk/gov/nationalarchives/Lambda.scala]: Do you need the trait here? The two case classes that extend it are doing different things and you're not doing any pattern matching on the `PreservicaVersion` trait. It might be simpler to remove it. What do you think?

**Review by MancunianSam — COMMENTED** [review]
Looks good, thanks. Just a few minor bits.

**Review by MancunianSam — APPROVED** [review]
